internal/resources/azure/mysql_flexible_server.go

```diff
@@ -0,0 +1,161 @@
+func (r *MySQLFlexibleServer) PopulateUsage(u *schema.UsageData) {
+	resources.PopulateArgsWithUsage(r, u)
+}
+
+// BuildResource builds a schema.Resource from a valid MySQLFlexibleServer struct.
+// This method is called after the resource is initialised by an IaC provider.
+// See providers folder for more information.
+func (r *MySQLFlexibleServer) BuildResource() *schema.Resource {
+	costComponents := []*schema.CostComponent{
+		r.computeCostComponent(),
+		r.storageCostComponent(),
+		r.iopsCostComponent(),
+		r.backupCostComponent(),
+	}
+
+	return &schema.Resource{
+		Name:           r.Address,
+		UsageSchema:    MySQLFlexibleServerUsageSchema,
+		CostComponents: costComponents,
+	}
+}
+
+// computeCostComponent returns a cost component for server compute requirements.
+func (r *MySQLFlexibleServer) computeCostComponent() *schema.CostComponent {
+	attrs := getFlexibleServerFilterAttributes(r.Tier, r.InstanceType, r.InstanceVersion)
+
+	// Eds series has two spaces in CPAPI hence '\s+'
+	productNameRegex := fmt.Sprintf("^Azure Database for MySQL Flexible Server %s\\s+%s", attrs.TierName, attrs.Series)
+
+	return &schema.CostComponent{
+		Name:           fmt.Sprintf("Compute (%s)", r.SKU),
+		Unit:           "hours",
+		UnitMultiplier: decimal.NewFromInt(1),
+		HourlyQuantity: decimalPtr(decimal.NewFromInt(1)),
+		ProductFilter: &schema.ProductFilter{
+			VendorName:    strPtr("azure"),
+			Region:        strPtr(r.Region),
+			Service:       strPtr("Azure Database for MySQL"),
+			ProductFamily: strPtr("Databases"),
+			AttributeFilters: []*schema.AttributeFilter{
+				{Key: "productName", ValueRegex: regexPtr(productNameRegex)},
+				{Key: "skuName", ValueRegex: regexPtr(fmt.Sprintf("^%s$", attrs.SKUName))},
+				{Key: "meterName", ValueRegex: regexPtr(fmt.Sprintf("^%s$", attrs.MeterName))},
+			},
+		},
+		PriceFilter: &schema.PriceFilter{
+			PurchaseOption: strPtr("Consumption"),
+		},
+	}
+}
+
+// storageCostComponent returns a cost component for server's storage.
+func (r *MySQLFlexibleServer) storageCostComponent() *schema.CostComponent {
+	var quantity *decimal.Decimal
+	if r.Storage > 0 {
+		quantity = decimalPtr(decimal.NewFromInt(r.Storage))
+	}
+
+	return &schema.CostComponent{
+		Name:            "Storage",
+		Unit:            "GB",
+		UnitMultiplier:  decimal.NewFromInt(1),
+		MonthlyQuantity: quantity,
+		ProductFilter: &schema.ProductFilter{
+			VendorName:    strPtr("azure"),
+			Region:        strPtr(r.Region),
+			Service:       strPtr("Azure Database for MySQL"),
+			ProductFamily: strPtr("Databases"),
+			AttributeFilters: []*schema.AttributeFilter{
+				{Key: "productName", Value: strPtr("Azure Database for MySQL Flexible Server Storage")},
+				{Key: "meterName", Value: strPtr("Storage Data Stored")},
+			},
+		},
+	}
+}
+
+// iopsCostComponent returns a cost component for additional IOPS.
+func (r *MySQLFlexibleServer) iopsCostComponent() *schema.CostComponent {
+	var quantity *decimal.Decimal
+	if r.IOPS > 0 {
```

Comment:
Not sure how to calculate the IOPS value:

1. Terraform has `storage.iops` https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/mysql_flexible_server#iops. Similar to it `storage_gb` represents the provisioned storage.
2. Pricing page mentions that each server includes free IOPS https://azure.microsoft.com/en-gb/pricing/details/mysql/flexible-server/:
   > the included IOPS of 3 IOPS/GB per total storage provisioned and free 300 IOPS available per server

   So, for example, the provisioned storage is 10GB, then the free IOPS would be 3*10 + 300 = 330. If the `storage.iops = 360` in TF resource, does it mean that the "additional IOPS" are 360 - 330 = 30 and we should show cost for these 30? NB: This formula doesn't apply for "Basic" burstable instances.
  There is another, but similar formula from [here](https://docs.microsoft.com/en-us/azure/mysql/flexible-server/concepts-compute-storage#iops) for "Complimentary IOPS": 
   > Complimentary IOPS are equal to MINIMUM("Max uncached disk throughput: IOPS/MBps" of compute size, 300 + storage provisioned in GiB * 3)

   Not sure if we can get this throughput value.

Or should we ignore the `storage.iops` value entirely and introduce a usage param`monthly_additional_iops` instead?

Reply:
> So, for example, the provisioned storage is 10GB

Provisioned storage seems to have a minimum of 20GB for all instance types when I try to launch in the Azure portal. So IOPS minimum is 360, which is free. Apart from that I think this formula makes sense.

> This formula doesn't apply for "Basic" burstable instances.

Yeah, in the Azure portal I'm seeing burstable instance types have a max IOPS listed. I don't know if we need to worry about representing this. We can test if the Terraform plan fails if it goes above that max IOPS. If so, then I think it's okay for us just to use the same formula for these.

Reply:
Updated.